src/com/amazon/corretto/crypto/provider/EvpKey.java

```diff
@@ -0,0 +1,195 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.amazon.corretto.crypto.provider;
+
+import java.io.IOException;
+import java.math.BigInteger;
+import java.security.AlgorithmParameters;
+import java.security.GeneralSecurityException;
+import java.security.Key;
+import java.security.MessageDigest;
+import java.security.PublicKey;
+import java.security.spec.AlgorithmParameterSpec;
+import java.util.Arrays;
+import java.util.Base64;
+import javax.security.auth.Destroyable;
+
+abstract class EvpKey implements Key, Destroyable {
+    private static final long serialVersionUID = 1;
+
+    protected final InternalKey internalKey;
+    protected final EvpKeyType type;
+    protected final boolean isPublicKey;
+    protected boolean ephemeral = false;
+
+    protected byte[] encoded;
+    protected Integer cachedHashCode;
```

Comment:
Is there a reason to make this an `Integer` not an `int`?

Reply:
my guess would be the `null` check on L147, which is needed to implement caching. from what i can tell, the [`Object.hashCode`'s javadoc](https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#hashCode()) doesn't put any constraints on the return value range over `int`, so presumably all `int`s are potentially valid `hashCode`s, precluding a flag value like -1 or something.

Reply:
This could also be done using a `bool isHashCodeCached` field in addition to an `int cachedHashCode`. I'm not sure how much of this is optimized away by the Java compiler, but I suspect having a separate `bool` field might avoid the performance cost of auto-boxing/unboxing the Integer each time it's used.

Reply:
My understanding is that auto-unboxing has gotten much better over the years. I think we can wait on this optimization until we have evidence it is needed.